vochain/cometbft.go

```diff
@@ -499,28 +503,140 @@ func (app *BaseApplication) ProcessProposal(_ context.Context,
+		}, nil
+	}
+	return &cometabcitypes.OfferSnapshotResponse{
+		Result: cometabcitypes.OFFER_SNAPSHOT_RESULT_REJECT,
+	}, nil
 }
 
-// LoadSnapshotChunk returns the response to a snapshot chunk loading request.
-func (*BaseApplication) LoadSnapshotChunk(context.Context,
-	*cometabcitypes.LoadSnapshotChunkRequest) (*cometabcitypes.LoadSnapshotChunkResponse, error) {
-	return &cometabcitypes.LoadSnapshotChunkResponse{}, nil
+// LoadSnapshotChunk provides cometbft with a snapshot chunk, during StateSync.
+//
+// cometbft will reject a len(Chunk) > 16M
+func (app *BaseApplication) LoadSnapshotChunk(_ context.Context,
+	req *cometabcitypes.LoadSnapshotChunkRequest,
+) (*cometabcitypes.LoadSnapshotChunkResponse, error) {
+	log.Debugw("CometBFT requests a Chunk from our Snapshot",
+		"height", req.Height, "format", req.Format, "chunk", req.Chunk)
+
+	buf, err := app.State.SliceSnapshotChunk(req.Height, req.Format, req.Chunk, app.stateSyncChunkSize)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &cometabcitypes.LoadSnapshotChunkResponse{
+		Chunk: buf,
+	}, nil
 }
 
-// ApplySnapshotChunk returns the response to a snapshot chunk applying request.
-func (*BaseApplication) ApplySnapshotChunk(context.Context,
-	*cometabcitypes.ApplySnapshotChunkRequest) (*cometabcitypes.ApplySnapshotChunkResponse, error) {
-	return &cometabcitypes.ApplySnapshotChunkResponse{}, nil
+// ApplySnapshotChunk applies a snapshot chunk provided by cometbft StateSync.
+//
+// cometbft will never pass a Chunk bigger than 16M
+func (app *BaseApplication) ApplySnapshotChunk(_ context.Context,
+	req *cometabcitypes.ApplySnapshotChunkRequest,
+) (*cometabcitypes.ApplySnapshotChunkResponse, error) {
+	log.Debugw("CometBFT provides us a Chunk",
+		"index", req.Index, "size", len(req.Chunk))
+	f, err := os.CreateTemp("", "SnapshotChunk")
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	if _, err := f.Write(req.Chunk); err != nil {
+		os.Remove(f.Name())
+		return nil, err
+	}
+
+	snaps.ChunkFilenames[req.Index] = f.Name()
+
+	// if we're missing chunks, just return
+	if len(snaps.ChunkFilenames) != int(snaps.Snap.GetChunks()) {
+		return &cometabcitypes.ApplySnapshotChunkResponse{
+			Result: cometabcitypes.APPLY_SNAPSHOT_CHUNK_RESULT_ACCEPT,
+		}, nil
+	}
+
+	// if we got here, all chunks are on disk
+	newState, err := app.State.JoinChunksAndRestoreSnapshot(snaps.ChunkFilenames, snaps.Snap.Height)
+	if err != nil {
+		log.Error(err)
+		return &cometabcitypes.ApplySnapshotChunkResponse{
+			Result: cometabcitypes.APPLY_SNAPSHOT_CHUNK_RESULT_REJECT_SNAPSHOT,
+		}, nil
+	}
+
+	newState.SetHeight(uint32(snaps.Snap.Height))
+
+	// TODO: dedup this from vochain/app.go NewBaseApplication
+	istc := ist.NewISTC(newState)
+	// Create the transaction handler for checking and processing transactions
+	transactionHandler := transaction.NewTransactionHandler(
+		newState,
+		istc,
+		filepath.Join(app.dataDir, "txHandler"),
+	)
+
+	app.State = newState
```

Comment:
This looks quite racy... app.State, app.Istc and app.TransactionHandler should probably be methods that hold a mutex (`lockSnapshotApply sync.Mutex`)